### PR TITLE
Test deb/rpm packages and publish them to apt/yum repos

### PR DESCRIPTION
**Pull request by thedewi**
This adds scripts to the archive with our linux packages that are used in the build and deploy:

- `test-linux-packages.sh` tests the `deb` and `rpm` packages in a variety of dockerized distros, to ensure `octo version` and `octo list-environments` work
- `publish-apt.sh` - creates a local repo based on our S3 repo, publishes our `deb`, pushes it back
- `publish-rpm.sh` - syncs the S3 repo to local, publishes our `rpm`, syncs back to S3
- `test-apt.sh` - tests the `apt` repo to ensure `octo version` and `octo list-environments` work
- `test-rpm.sh` - tests the `rpm` repo with yum to ensure `octo version` and `octo list-environments` work

Please review the contents of the scripts.

### CODE REVIEW

File: BuildAssets/repos/publish-apt.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+# Required env vars:
+#export AWS_ACCESS_KEY_ID=$(get_octopusvariable "OctopusToolsAwsAccount.AccessKey")
```
Comment (matt-richardson) [inline BuildAssets/repos/publish-apt.sh]: future improvement: we can make this fail if the env var isn't there

File: BuildAssets/repos/publish-apt.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#export S3_PUBLISH_ENDPOINT=$(get_octopusvariable "Publish.APT.S3.TargetBucket")
+#export GPG_PRIVATEKEY=$(get_octopusvariable "Publish.APT.GPG.PrivateKey")
+#export GPG_PASSPHRASE=$(get_octopusvariable "Publish.APT.GPG.PassPhrase")
+
+echo "Importing private key"
+echo "$GPG_PRIVATEKEY" | gpg1 --batch --import 2>&1 || exit
+curl -s "https://s3.amazonaws.com/$S3_PUBLISH_ENDPOINT/public.key" | gpg1 --no-default-keyring --keyring trustedkeys.gpg --import 2>&1 || exit
+
+echo "Configuring S3 bucket"
+#aws s3 mb "s3://$S3_PUBLISH_ENDPOINT" || exit 1
```
Comment (matt-richardson) [inline BuildAssets/repos/publish-apt.sh]: future improvement: let's figure out how to make this work between tentacle / octocli

File: BuildAssets/repos/publish-apt.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#aws s3api wait bucket-exists --bucket "$S3_PUBLISH_ENDPOINT" || exit 1
+#aws s3 sync ./apt-content "s3://$S3_PUBLISH_ENDPOINT" --acl public-read || exit 1
+aptly config show 2>/dev/null | jq '.S3PublishEndpoints[env.S3_PUBLISH_ENDPOINT] = {"region": "us-east-1", "bucket": env.S3_PUBLISH_ENDPOINT, "acl": "public-read"}' > ~/.aptly.conf.new || exit
+mv ~/.aptly.conf.new ~/.aptly.conf || exit
+
+echo "Creating APT repo"
+aptly repo create -distribution=stretch -component=main octopus || exit
```
Comment (matt-richardson) [inline BuildAssets/repos/publish-apt.sh]: future: (as you've mentioned) we need to make this handle different distributions

File: BuildAssets/repos/publish-rpm.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+# Required env vars:
+#export AWS_ACCESS_KEY_ID=$(get_octopusvariable "OctopusToolsAwsAccount.AccessKey")
```
Comment (matt-richardson) [inline BuildAssets/repos/publish-rpm.sh]: future improvement: we can make this fail if the env var isn't there

File: BuildAssets/repos/publish-rpm.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#aws s3api wait bucket-exists --bucket ${S3_PUBLISH_ENDPOINT} || exit 1
+#aws s3 sync ./rpm-content "s3://${S3_PUBLISH_ENDPOINT}" --acl public-read || exit 1
+echo -e "[octopuscli]
+name=Octopus CLI
+baseurl=https://s3.amazonaws.com/$S3_PUBLISH_ENDPOINT/\$basearch/
+enabled=1
+gpgkey=https://s3.amazonaws.com/$S3_PUBLISH_ENDPOINT/public.key
+gpgcheck=0
+" > octopuscli.repo
+aws s3 cp octopuscli.repo "s3://${S3_PUBLISH_ENDPOINT}/octopuscli.repo" --acl public-read || exit
```
Comment (matt-richardson) [inline BuildAssets/repos/publish-rpm.sh]: need to confirm this matches our octofront page

File: BuildAssets/test-linux-packages.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+# This script requires environment variables to run:
+#   OCTOPUS_CLI_SERVER, OCTOPUS_CLI_API_KEY, OCTOPUS_SPACE, OCTOPUS_EXPECT_ENV
+# You must also supply either:
+#   TEST_QUICK (to skip some distributions)
+# or
+#   REDHAT_SUBSCRIPTION_USERNAME, REDHAT_SUBSCRIPTION_PASSWORD (to support RHEL tests)
+
+TEST_DEB_SH='
+  # Configure apt
+  export DEBIAN_FRONTEND=noninteractive
```
Comment (matt-richardson) [inline BuildAssets/test-linux-packages.sh]: I've read this is discouraged... https://docs.docker.com/engine/faq/#why-is-debian_frontendnoninteractive-discouraged-in-dockerfiles
Is that relevant to us?

File: BuildAssets/test-linux-packages.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  apt-get update --quiet 2 || exit
+  apt-get install --no-install-recommends --yes apt-utils >/dev/null 2>&1 || exit # silence debconf warnings
+
+  # Install octo
+  dpkg -i /pkgs/octopuscli*.deb >/dev/null 2>&1 # Silenced and expected to fail
```
Comment (matt-richardson) [inline BuildAssets/test-linux-packages.sh]: can you expand why it's expected to fail?

File: BuildAssets/test-linux-packages.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+# This script requires environment variables to run:
+#   OCTOPUS_CLI_SERVER, OCTOPUS_CLI_API_KEY, OCTOPUS_SPACE, OCTOPUS_EXPECT_ENV
+# You must also supply either:
+#   TEST_QUICK (to skip some distributions)
+# or
+#   REDHAT_SUBSCRIPTION_USERNAME, REDHAT_SUBSCRIPTION_PASSWORD (to support RHEL tests)
+
+TEST_DEB_SH='
```
Comment (matt-richardson) [inline BuildAssets/test-linux-packages.sh]: Any reason you didn't do these as separate files?

File: BuildAssets/repos/publish-rpm.sh (hunk 5 above)
Comment (thedewi) [reply BuildAssets/repos/publish-rpm.sh]: It does.

File: BuildAssets/test-linux-packages.sh (hunk 6 above)
Comment (thedewi) [reply BuildAssets/test-linux-packages.sh]: Interesting. I think it will be okay when we're not building an image or will ever run interactively.

File: BuildAssets/test-linux-packages.sh (hunk 7 above)
Comment (thedewi) [reply BuildAssets/test-linux-packages.sh]: Done.

File: BuildAssets/test-linux-packages.sh (hunk 8 above)
Comment (thedewi) [reply BuildAssets/test-linux-packages.sh]: No strong reason. I think the self-containedness makes it robust and easier to understand but those blocks are a little harder to read, I admit.

File: BuildAssets/repos/publish-apt.sh (hunk 1 above)
Comment (thedewi) [reply BuildAssets/repos/publish-apt.sh]: Addressed in #55.

File: BuildAssets/repos/publish-apt.sh (hunk 2 above)
Comment (thedewi) [reply BuildAssets/repos/publish-apt.sh]: Addressed in #55.

File: BuildAssets/repos/publish-apt.sh (hunk 3 above)
Comment (thedewi) [reply BuildAssets/repos/publish-apt.sh]: Addressed in #55.

File: BuildAssets/repos/publish-rpm.sh (hunk 4 above)
Comment (thedewi) [reply BuildAssets/repos/publish-rpm.sh]: Addressed in #55.
